File: Backend/src/main/java/com/webbandogiadung/backend/service/OrderService.java
package com.webbandogiadung.backend.service;

import com.webbandogiadung.backend.dto.CreateOrderRequest;
import com.webbandogiadung.backend.dto.OrderItemRequest;
import com.webbandogiadung.backend.dto.OrderResponse;
import com.webbandogiadung.backend.entity.CustomerOrder;
import com.webbandogiadung.backend.entity.OrderItem;
import com.webbandogiadung.backend.entity.Product;
import com.webbandogiadung.backend.exception.BadRequestException;
import com.webbandogiadung.backend.exception.ResourceNotFoundException;
import com.webbandogiadung.backend.repository.OrderRepository;
import com.webbandogiadung.backend.repository.ProductRepository;
import java.math.BigDecimal;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class OrderService {

    private final OrderRepository orderRepository;
    private final ProductRepository productRepository;
    private final OrderMapper orderMapper;

    public OrderService(OrderRepository orderRepository, ProductRepository productRepository, OrderMapper orderMapper) {
        this.orderRepository = orderRepository;
        this.productRepository = productRepository;
        this.orderMapper = orderMapper;
    }

    @Transactional
    public OrderResponse createOrder(CreateOrderRequest request) {
        CustomerOrder order = new CustomerOrder();
        order.setEmail(request.email());
        order.setFirstName(request.firstName());
        order.setLastName(request.lastName());
        order.setPhone(request.phone());
        order.setAddress(request.address());
        order.setApartment(request.apartment());
        order.setCity(request.city());
        order.setState(request.state());
        order.setZip(request.zip());

        BigDecimal subtotal = BigDecimal.ZERO;
        for (OrderItemRequest itemRequest : request.items()) {
            Product product = productRepository.findById(itemRequest.productId())
                    .orElseThrow(() -> new ResourceNotFoundException(
                            "Khong tim thay san pham id=" + itemRequest.productId()));

            if (itemRequest.quantity() > product.getStock()) {
                throw new BadRequestException("San pham " + product.getName() + " chi con " + product.getStock());
            }

            OrderItem item = new OrderItem(product.getId(), product.getName(), product.getPrice(), itemRequest.quantity());
            order.addItem(item);
            subtotal = subtotal.add(item.getLineTotal());
        }

        order.setSubtotal(subtotal);
        order.setShippingFee(BigDecimal.ZERO);
        order.setTotal(subtotal.add(order.getShippingFee()));

        return orderMapper.toResponse(orderRepository.save(order));
    }

    @Transactional(readOnly = true)
    public OrderResponse getOrder(Long id) {
        return orderRepository.findWithItemsById(id)
                .map(orderMapper::toResponse)
                .orElseThrow(() -> new ResourceNotFoundException("Khong tim thay don hang id=" + id));
    }
}

File: Backend/Services/OrderService.cs
using Backend.Dtos;
using Backend.Enums;
using Backend.Exceptions;
using Backend.Models;
using Backend.Repositories;

namespace Backend.Services;

public sealed class OrderService(
    IOrderRepository orderRepository,
    IProductRepository productRepository,
    CouponService couponService)
{
    public IReadOnlyList<OrderResponse> GetOrders() =>
        orderRepository.GetAll().Select(ToResponse).ToList();

    public OrderResponse GetOrder(long id)
    {
        var order = orderRepository.GetById(id)
            ?? throw new ResourceNotFoundException($"Khong tim thay don hang id={id}");

        return ToResponse(order);
    }

    public OrderResponse CreateOrder(CreateOrderRequest request)
    {
        var order = new CustomerOrder
        {
            Email = request.Email.Trim(),
            FirstName = request.FirstName.Trim(),
            LastName = request.LastName.Trim(),
            Phone = request.Phone?.Trim(),
            Address = request.Address.Trim(),
            Apartment = request.Apartment?.Trim(),
            City = request.City.Trim(),
            State = request.State?.Trim(),
            Zip = request.Zip?.Trim(),
            PaymentMethod = request.PaymentMethod ?? PaymentMethod.Cod,
            CreatedAt = DateTime.UtcNow
        };

        foreach (var itemRequest in request.Items)
        {
            var product = productRepository.GetById(itemRequest.ProductId)
                ?? throw new ResourceNotFoundException($"Khong tim thay san pham id={itemRequest.ProductId}");

            if (itemRequest.Quantity > product.Stock)
            {
                throw new BadRequestException($"San pham {product.Name} chi con {product.Stock}");
            }

            order.Items.Add(new OrderItem
            {
                ProductId = product.Id,
                ProductName = product.Name,
                UnitPrice = product.Price,
                Quantity = itemRequest.Quantity
            });
        }

        order.Subtotal = order.Items.Sum(item => item.LineTotal);
        order.ShippingFee = 0;

        if (!string.IsNullOrWhiteSpace(request.CouponCode))
        {
            var validation = couponService.ValidateCoupon(request.CouponCode, order.Subtotal);
            order.CouponCode = validation.Code;
            order.DiscountAmount = validation.DiscountAmount;
            couponService.IncrementUsage(validation.Code);
        }

        order.Total = order.Subtotal + order.ShippingFee - order.DiscountAmount;
        order.PaymentStatus = order.PaymentMethod == PaymentMethod.Cod ? PaymentStatus.Unpaid : PaymentStatus.Pending;
        order.PaidAt = null;

        return ToResponse(orderRepository.Add(order));
    }

    public OrderResponse UpdateStatus(long id, OrderStatus status)
    {
        var order = orderRepository.UpdateStatus(id, status)
            ?? throw new ResourceNotFoundException($"Khong tim thay don hang id={id}");

        return ToResponse(order);
    }

    private static OrderResponse ToResponse(CustomerOrder order) => new(
        order.Id,
        order.Email,
        $"{order.FirstName} {order.LastName}",
        order.Address,
        order.City,
        order.Subtotal,
        order.ShippingFee,
        order.DiscountAmount,
        order.CouponCode,
        order.Total,
        order.Status,
        order.PaymentMethod,
        order.PaymentStatus,
        order.PaymentTransactionId,
        order.CreatedAt,
        order.PaidAt,
        order.CancelledAt,
        order.CancelReason,
        order.Items.Select(item => new OrderItemResponse(
            item.ProductId,
            item.ProductName,
            item.UnitPrice,
            item.Quantity,
            item.LineTotal)).ToList());
}
